In [1]:
# =============================================================================
# STEP 1: LOAD PROCESSED DATASET
# =============================================================================

import pandas as pd
import numpy as np


# Load cleaned dataset
df = pd.read_csv(
    "../../data/processed/breast_cancer_cleaned.csv"
)


print("=" * 70)
print("PROCESSED DATASET")
print("=" * 70)

print(f"Rows    : {df.shape[0]}")
print(f"Columns : {df.shape[1]}")


display(df.head())

PROCESSED DATASET
Rows    : 569
Columns : 31


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension,target
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,0
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,0
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,0
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,0
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,0


In [2]:
# =============================================================================
# STEP 2: FEATURE AND TARGET SEPARATION
# =============================================================================


target_column = "target"


X = df.drop(
    target_column,
    axis=1
)


y = df[target_column]



print("=" * 70)
print("FEATURE MATRIX (X)")
print("=" * 70)

print(f"Rows    : {X.shape[0]}")
print(f"Columns : {X.shape[1]}")


print("\n" + "=" * 70)
print("TARGET VARIABLE (y)")
print("=" * 70)

print(f"Rows : {y.shape[0]}")

display(y.head())

FEATURE MATRIX (X)
Rows    : 569
Columns : 30

TARGET VARIABLE (y)
Rows : 569


0    0
1    0
2    0
3    0
4    0
Name: target, dtype: int64

In [3]:
# =============================================================================
# STEP 3: VERIFY TARGET DISTRIBUTION
# =============================================================================


print("=" * 70)
print("TARGET DISTRIBUTION")
print("=" * 70)


display(
    y.value_counts()
)


print("\nClass Percentage")


display(
    (y.value_counts(normalize=True)*100)
    .round(2)
)

TARGET DISTRIBUTION


target
1    357
0    212
Name: count, dtype: int64


Class Percentage


target
1    62.74
0    37.26
Name: proportion, dtype: float64

In [4]:
# =============================================================================
# STEP 4: FEATURE TYPE CHECK
# =============================================================================


print("=" * 70)
print("FEATURE DATA TYPES")
print("=" * 70)


display(
    X.dtypes.to_frame(
        name="Data Type"
    )
)

FEATURE DATA TYPES


,Data Type
mean radius,float64
mean texture,float64
mean perimeter,float64
mean area,float64
mean smoothness,float64
mean compactness,float64
mean concavity,float64
mean concave points,float64
mean symmetry,float64
mean fractal dimension,float64


In [5]:
# =============================================================================
# STEP 3: TRAIN-TEST SPLIT
# =============================================================================


from sklearn.model_selection import train_test_split


print("=" * 70)
print("TRAIN-TEST SPLIT")
print("=" * 70)


X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)


print("Training Data")
print("-" * 70)

print(f"X_train shape : {X_train.shape}")
print(f"y_train shape : {y_train.shape}")


print("\nTesting Data")
print("-" * 70)

print(f"X_test shape  : {X_test.shape}")
print(f"y_test shape  : {y_test.shape}")

TRAIN-TEST SPLIT
Training Data
----------------------------------------------------------------------
X_train shape : (455, 30)
y_train shape : (455,)

Testing Data
----------------------------------------------------------------------
X_test shape  : (114, 30)
y_test shape  : (114,)


In [6]:
# =============================================================================
# CLASS DISTRIBUTION AFTER SPLIT
# =============================================================================


print("=" * 70)
print("TRAINING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_train.value_counts(normalize=True)*100)
    .round(2)
)



print("\n" + "=" * 70)
print("TESTING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_test.value_counts(normalize=True)*100)
    .round(2)
)

TRAINING TARGET DISTRIBUTION


target
1    62.64
0    37.36
Name: proportion, dtype: float64


TESTING TARGET DISTRIBUTION


target
1    63.16
0    36.84
Name: proportion, dtype: float64

In [7]:
# =============================================================================
# STEP 4: FEATURE SCALING
# =============================================================================


from sklearn.preprocessing import StandardScaler


print("=" * 70)
print("FEATURE SCALING USING STANDARD SCALER")
print("=" * 70)


# Initialize scaler
scaler = StandardScaler()


# Fit only on training data
X_train_scaled = scaler.fit_transform(X_train)


# Transform test data using same scaler
X_test_scaled = scaler.transform(X_test)



print("Scaling completed successfully")


print("\n" + "=" * 70)
print("SHAPE AFTER SCALING")
print("=" * 70)


print(f"X_train_scaled shape : {X_train_scaled.shape}")
print(f"X_test_scaled shape  : {X_test_scaled.shape}")

FEATURE SCALING USING STANDARD SCALER
Scaling completed successfully

SHAPE AFTER SCALING
X_train_scaled shape : (455, 30)
X_test_scaled shape  : (114, 30)


In [8]:
# =============================================================================
# CONVERT SCALED DATA BACK TO DATAFRAME
# =============================================================================


X_train_scaled = pd.DataFrame(
    X_train_scaled,
    columns=X_train.columns
)


X_test_scaled = pd.DataFrame(
    X_test_scaled,
    columns=X_test.columns
)



print("=" * 70)
print("SCALED TRAIN DATA PREVIEW")
print("=" * 70)


display(X_train_scaled.head())

SCALED TRAIN DATA PREVIEW


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst radius,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension
0,-1.072001,-0.658425,-1.088080,-0.939274,-0.135940,-1.008718,-0.968359,-1.102032,0.281062,-0.113231,...,-1.034094,-0.623497,-1.070773,-0.876534,-0.169982,-1.038836,-1.078995,-1.350527,-0.352658,-0.541380
1,1.748743,0.066502,1.751157,1.745559,1.274468,0.842288,1.519852,1.994664,-0.293045,-0.320180,...,1.228342,-0.092833,1.187467,1.104386,1.517001,0.249655,1.178594,1.549916,0.191078,-0.173739
2,-0.974734,-0.931124,-0.997709,-0.867589,-0.613515,-1.138154,-1.092292,-1.243358,0.434395,-0.429247,...,-0.973231,-1.036772,-1.008044,-0.834168,-1.097823,-1.167260,-1.282241,-1.707442,-0.307734,-1.213033
3,-0.145103,-1.215186,-0.123013,-0.253192,0.664482,0.286762,-0.129729,-0.098605,0.555635,0.029395,...,-0.251266,-1.369643,-0.166633,-0.330292,0.234006,0.096874,-0.087521,-0.344838,0.242198,-0.118266
4,-0.771617,-0.081211,-0.803700,-0.732927,-0.672282,-1.006099,-0.798502,-0.684484,0.737495,-0.457213,...,-0.801135,0.079230,-0.824381,-0.741830,-0.911367,-0.984612,-0.933190,-0.777604,0.555118,-0.761639


In [9]:
# =============================================================================
# SCALING VERIFICATION
# =============================================================================


print("=" * 70)
print("MEAN AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.mean().round(2)
)



print("\n" + "=" * 70)
print("STANDARD DEVIATION AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.std().round(2)
)

MEAN AFTER SCALING


mean radius               -0.0
mean texture               0.0
mean perimeter            -0.0
mean area                 -0.0
mean smoothness            0.0
mean compactness          -0.0
mean concavity            -0.0
mean concave points        0.0
mean symmetry             -0.0
mean fractal dimension     0.0
radius error              -0.0
texture error             -0.0
perimeter error            0.0
area error                 0.0
smoothness error          -0.0
compactness error         -0.0
concavity error            0.0
concave points error      -0.0
symmetry error             0.0
fractal dimension error    0.0
worst radius               0.0
worst texture             -0.0
worst perimeter           -0.0
worst area                 0.0
worst smoothness           0.0
worst compactness         -0.0
worst concavity            0.0
worst concave points      -0.0
worst symmetry            -0.0
worst fractal dimension    0.0
dtype: float64


STANDARD DEVIATION AFTER SCALING


mean radius                1.0
mean texture               1.0
mean perimeter             1.0
mean area                  1.0
mean smoothness            1.0
mean compactness           1.0
mean concavity             1.0
mean concave points        1.0
mean symmetry              1.0
mean fractal dimension     1.0
radius error               1.0
texture error              1.0
perimeter error            1.0
area error                 1.0
smoothness error           1.0
compactness error          1.0
concavity error            1.0
concave points error       1.0
symmetry error             1.0
fractal dimension error    1.0
worst radius               1.0
worst texture              1.0
worst perimeter            1.0
worst area                 1.0
worst smoothness           1.0
worst compactness          1.0
worst concavity            1.0
worst concave points       1.0
worst symmetry             1.0
worst fractal dimension    1.0
dtype: float64

# Random Forest

# 1. Baseline

In [10]:
# =============================================================================
# RANDOM FOREST BASELINE FEATURE SELECTION
# TOP-K = 5, 10, 15, 20
# =============================================================================

from sklearn.ensemble import RandomForestClassifier
import pandas as pd


# =============================================================================
# FIT RANDOM FOREST ON ALL TRAINING FEATURES
# =============================================================================

rf_selector = RandomForestClassifier(
    n_estimators=100,
    class_weight="balanced",
    random_state=42,
    n_jobs=-1
)


rf_selector.fit(
    X_train_scaled,
    y_train
)


# =============================================================================
# CREATE FEATURE IMPORTANCE DATAFRAME
# =============================================================================

rf_scores = pd.DataFrame({

    "Feature": X_train_scaled.columns,

    "RF Importance": rf_selector.feature_importances_

})


# =============================================================================
# SORT FEATURES BY IMPORTANCE
# =============================================================================

rf_scores = rf_scores.sort_values(
    by="RF Importance",
    ascending=False
).reset_index(drop=True)


# =============================================================================
# TOP-K VALUES
# =============================================================================

k_values = [5, 10, 15, 20]

rf_results = {}


# =============================================================================
# SELECT TOP-K FEATURES
# =============================================================================

for k in k_values:

    top_features = (
        rf_scores
        .head(k)["Feature"]
        .tolist()
    )


    # -------------------------------------------------------------------------
    # Boolean mask for selecting columns
    # -------------------------------------------------------------------------

    mask = X_train_scaled.columns.isin(
        top_features
    )


    X_train_selected = X_train_scaled.loc[
        :, mask
    ]

    X_test_selected = X_test_scaled.loc[
        :, mask
    ]


    # -------------------------------------------------------------------------
    # Store results
    # -------------------------------------------------------------------------

    rf_results[k] = {

        "features": top_features,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "scores": rf_scores.head(k)

    }


    # -------------------------------------------------------------------------
    # Display results
    # -------------------------------------------------------------------------

    print("=" * 70)
    print(f"TOP {k} RANDOM FOREST FEATURES")
    print("=" * 70)

    print("\nSelected Features:")
    print(top_features)

    print("\nFeature Importance:")
    print(rf_scores.head(k))


# =============================================================================
# RANDOM FOREST BASELINE FEATURE SELECTION COMPLETED
# =============================================================================

TOP 5 RANDOM FOREST FEATURES

Selected Features:
['worst area', 'worst concave points', 'mean concave points', 'worst perimeter', 'worst radius']

Feature Importance:
                Feature  RF Importance
0            worst area       0.137862
1  worst concave points       0.131157
2   mean concave points       0.116819
3       worst perimeter       0.081788
4          worst radius       0.076800
TOP 10 RANDOM FOREST FEATURES

Selected Features:
['worst area', 'worst concave points', 'mean concave points', 'worst perimeter', 'worst radius', 'mean concavity', 'mean perimeter', 'mean area', 'worst concavity', 'mean radius']

Feature Importance:
                Feature  RF Importance
0            worst area       0.137862
1  worst concave points       0.131157
2   mean concave points       0.116819
3       worst perimeter       0.081788
4          worst radius       0.076800
5        mean concavity       0.068669
6        mean perimeter       0.065251
7             mean area       0.0405

In [11]:
# =============================================================================
# BASELINE MODELS
# =============================================================================

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier


models = {

    "Logistic Regression": LogisticRegression(
        max_iter=1000,
        class_weight="balanced",
        random_state=42
    ),

    "Random Forest": RandomForestClassifier(
        n_estimators=100,
        class_weight="balanced",
        random_state=42
    ),

    "XGBoost": XGBClassifier(
        random_state=42,
        eval_metric="logloss",
        scale_pos_weight=(
            (y_train == 0).sum() /
            (y_train == 1).sum()
        )
    )
}

In [12]:
# =============================================================================
# RANDOM FOREST FEATURE SELECTION — MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)


rf_model_results = []


for k in k_values:

    Xtr = rf_results[k]["X_train"]

    Xts = rf_results[k]["X_test"]

    features = ", ".join(
        rf_results[k]["features"]
    )


    for model_name, model in models.items():

        model.fit(
            Xtr,
            y_train
        )


        y_pred = model.predict(
            Xts
        )


        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        rf_model_results.append({

            "Method": "Random Forest",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "Recall": recall_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )
        })


rf_results_df = pd.DataFrame(
    rf_model_results
)


display(rf_results_df)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,Random Forest,5,Logistic Regression,"worst area, worst concave points, mean concave...",0.938596,0.985075,0.916667,0.949640,0.990741
1,Random Forest,5,Random Forest,"worst area, worst concave points, mean concave...",0.938596,0.971014,0.930556,0.950355,0.986772
2,Random Forest,5,XGBoost,"worst area, worst concave points, mean concave...",0.921053,0.943662,0.930556,0.937063,0.982474
3,Random Forest,10,Logistic Regression,"worst area, worst concave points, mean concave...",0.912281,0.969697,0.888889,0.927536,0.991402
4,Random Forest,10,Random Forest,"worst area, worst concave points, mean concave...",0.956140,0.958904,0.972222,0.965517,0.990079
5,Random Forest,10,XGBoost,"worst area, worst concave points, mean concave...",0.973684,0.972603,0.986111,0.979310,0.992394
6,Random Forest,15,Logistic Regression,"worst area, worst concave points, mean concave...",0.938596,0.957746,0.944444,0.951049,0.994048
7,Random Forest,15,Random Forest,"worst area, worst concave points, mean concave...",0.947368,0.958333,0.958333,0.958333,0.992725
8,Random Forest,15,XGBoost,"worst area, worst concave points, mean concave...",0.947368,0.958333,0.958333,0.958333,0.995040
9,Random Forest,20,Logistic Regression,"worst area, worst concave points, mean concave...",0.956140,0.985507,0.944444,0.964539,0.995701


In [13]:
rf_results_df.to_csv(
    "../../../results/breast_cancer/baseline_results/rf_baseline_results.csv",
    index=False
)

## 2. DODA

In [14]:
import sys
import subprocess

subprocess.check_call([
    sys.executable,
    "-m",
    "pip",
    "install",
    "--upgrade",
    "--force-reinstall",
    "--no-cache-dir",
    "git+https://github.com/anandha-3679/DODA.git"
])

0

In [15]:
import sys
import doda
from doda import DODASelector
import inspect

print("Python:", sys.executable)
print("DODA:", doda.__file__)
print("Selector:", inspect.getfile(DODASelector))

Python: c:\Users\johnm\msc_research\.venv\Scripts\python.exe
DODA: c:\Users\johnm\msc_research\.venv\Lib\site-packages\doda\__init__.py
Selector: c:\Users\johnm\msc_research\.venv\Lib\site-packages\doda\selector.py


In [16]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

from doda import DODASelector

from doda.adapters import SklearnAdapter

from doda.knowledge.providers import JSONProvider

from doda.fusion import HadamardFusion

In [17]:
# =============================================================================
# RANDOM FOREST + DODA FEATURE SELECTION
# =============================================================================

import pandas as pd
from sklearn.ensemble import RandomForestClassifier


# -------------------------------------------------------------------------
# Top-K values
# -------------------------------------------------------------------------

k_values = [5, 10, 15, 20]


# -------------------------------------------------------------------------
# Clinical knowledge provider
# -------------------------------------------------------------------------

provider = JSONProvider(
    "../../../config/clinical_weights/breast_cancer.json"
)


# -------------------------------------------------------------------------
# Fusion method
# -------------------------------------------------------------------------

fusion = HadamardFusion()


# -------------------------------------------------------------------------
# Store DODA results
# -------------------------------------------------------------------------

rf_doda_results = {}


# =============================================================================
# RUN DODA FOR EACH TOP-K
# =============================================================================

for k in k_values:

    print("\n" + "=" * 70)
    print(f"RANDOM FOREST + DODA \u2014 TOP {k} FEATURES")
    print("=" * 70)


    # ---------------------------------------------------------------------
    # Random Forest operator
    # ---------------------------------------------------------------------

    rf = SklearnAdapter(
        RandomForestClassifier(
            n_estimators=100,
            class_weight="balanced",
            random_state=42,
            n_jobs=-1
        )
    )


    # ---------------------------------------------------------------------
    # DODA selector
    # ---------------------------------------------------------------------

    selector = DODASelector(
        operators=[rf],
        provider=provider,
        fusion=fusion,
        top_k=k
    )


    # ---------------------------------------------------------------------
    # Fit DODA on TRAINING data
    # ---------------------------------------------------------------------

    X_train_selected = selector.fit_transform(
        X_train_scaled,
        y_train
    )


    # ---------------------------------------------------------------------
    # Transform TEST data
    # ---------------------------------------------------------------------

    X_test_selected = selector.transform(
        X_test_scaled
    )


    # ---------------------------------------------------------------------
    # Selected features
    # ---------------------------------------------------------------------

    selected_features = selector.get_selected_features()


    print("\nSelected Features:")
    print(selected_features)


    # ---------------------------------------------------------------------
    # Store everything needed for later evaluation
    # ---------------------------------------------------------------------

    rf_doda_results[k] = {

        "selector": selector,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "features": selected_features,

        "raw_math_scores": selector.raw_math_scores_,

        "math_scores": selector.math_scores_,

        "clinical_weights": selector.clinical_weights_,

        "final_scores": selector.final_scores_
    }


print("\n" + "=" * 70)
print("RANDOM FOREST + DODA FEATURE SELECTION COMPLETED")
print("=" * 70)


RANDOM FOREST + DODA — TOP 5 FEATURES
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x00000258E3A49160>
DODA Pipeline Started
Running: RandomForestClassifier
DODA feature order:
['mean radius', 'mean texture', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean concave points', 'mean symmetry', 'mean fractal dimension']
Number of features: 30
Number of feature importances: 30
Operator Scores:
{'RandomForestClassifier_1': {'mean radius': np.float64(0.037878571075900605), 'mean texture': np.float64(0.011584227153273762), 'mean perimeter': np.float64(0.06525109582346444), 'mean area': np.float64(0.04052144874956467), 'mean smoothness': np.float64(0.007004956117366906), 'mean compactness': np.float64(0.012885022993264956), 'mean concavity': np.float64(0.06866896338978329), 'mean concave points': np.float64(0.1168193189507702), 'mean symmetry': np.float64(0.003998715308884734), 'mean fractal dimens

In [18]:
# CELL 2
# =============================================================================
# INSPECT MATHEMATICAL SCORES
# =============================================================================

k = 10

selector = rf_doda_results[k]["selector"]

# -----------------------------------------------------------------------------
# Raw mathematical scores
# -----------------------------------------------------------------------------

raw_math_scores = pd.DataFrame(
    selector.raw_math_scores_.items(),
    columns=[
        "Feature",
        "Raw Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Normalized mathematical scores
# -----------------------------------------------------------------------------

normalized_math_scores = pd.DataFrame(
    selector.math_scores_.items(),
    columns=[
        "Feature",
        "Normalized Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Combine both
# -----------------------------------------------------------------------------

math_scores = raw_math_scores.merge(
    normalized_math_scores,
    on="Feature"
)

math_scores = math_scores.sort_values(
    by="Normalized Math Score",
    ascending=False
)

print("=" * 70)
print("MATHEMATICAL SCORES")
print("=" * 70)

display(math_scores)

MATHEMATICAL SCORES


,Feature,Raw Math Score,Normalized Math Score
23,worst area,0.137862,1.000000
27,worst concave points,0.131157,0.951366
7,mean concave points,0.116819,0.847365
22,worst perimeter,0.081788,0.593258
20,worst radius,0.076800,0.557076
6,mean concavity,0.068669,0.498100
2,mean perimeter,0.065251,0.473308
3,mean area,0.040521,0.293928
26,worst concavity,0.040485,0.293665
0,mean radius,0.037879,0.274757


In [19]:
# CELL 3
# =============================================================================
# INSPECT CLINICAL WEIGHTS
# =============================================================================

clinical_weights = pd.DataFrame(
    selector.clinical_weights_.items(),
    columns=[
        "Feature",
        "Clinical Weight"
    ]
)

clinical_weights = clinical_weights.sort_values(
    by="Clinical Weight",
    ascending=False
)

print("=" * 70)
print("CLINICAL WEIGHTS")
print("=" * 70)

display(clinical_weights)

CLINICAL WEIGHTS


,Feature,Clinical Weight
27,worst concave points,1.0
26,worst concavity,1.0
22,worst perimeter,1.0
23,worst area,1.0
20,worst radius,1.0
0,mean radius,0.9
3,mean area,0.9
6,mean concavity,0.9
7,mean concave points,0.9
2,mean perimeter,0.9


In [20]:
# CELL 4
# =============================================================================
# INSPECT FINAL DODA SCORES
# =============================================================================

final_scores = pd.DataFrame(
    selector.final_scores_.items(),
    columns=[
        "Feature",
        "Final DODA Score"
    ]
)

final_scores = final_scores.sort_values(
    by="Final DODA Score",
    ascending=False
)

print("=" * 70)
print("FINAL DODA SCORES")
print("=" * 70)

display(final_scores)

FINAL DODA SCORES


,Feature,Final DODA Score
23,worst area,1.000000
27,worst concave points,0.951366
7,mean concave points,0.762629
22,worst perimeter,0.593258
20,worst radius,0.557076
6,mean concavity,0.448290
2,mean perimeter,0.425977
26,worst concavity,0.293665
3,mean area,0.264535
0,mean radius,0.247282


In [21]:
selector.get_selected_features()

['worst area',
 'worst concave points',
 'mean concave points',
 'worst perimeter',
 'worst radius',
 'mean concavity',
 'mean perimeter',
 'worst concavity',
 'mean area',
 'mean radius']

In [22]:
comparison = selector.compare_scores()

comparison

,Feature,Math Rank,Normalized Math Score,Clinical Weight,DODA Rank,Final Score,Rank Change
0,worst area,1,1.0000,1.0,1,1.0000,0
1,worst concave points,2,0.9514,1.0,2,0.9514,0
2,mean concave points,3,0.8474,0.9,3,0.7626,0
3,worst perimeter,4,0.5933,1.0,4,0.5933,0
4,worst radius,5,0.5571,1.0,5,0.5571,0
5,mean concavity,6,0.4981,0.9,6,0.4483,0
6,mean perimeter,7,0.4733,0.9,7,0.4260,0
7,worst concavity,9,0.2937,1.0,8,0.2937,1
8,mean area,8,0.2939,0.9,9,0.2645,-1
9,mean radius,10,0.2748,0.9,10,0.2473,0


In [23]:
# =============================================================================
# RANDOM FOREST + DODA MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)


rf_doda_model_results = []


for k in k_values:

    Xtr = rf_doda_results[k]["X_train"]

    Xts = rf_doda_results[k]["X_test"]

    features = ", ".join(
        rf_doda_results[k]["features"]
    )


    for model_name, model in models.items():

        model.fit(
            Xtr,
            y_train
        )


        y_pred = model.predict(
            Xts
        )


        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        rf_doda_model_results.append({

            "Method": "RF + DODA",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "Recall": recall_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )
        })


rf_doda_results_df = pd.DataFrame(
    rf_doda_model_results
)


display(rf_doda_results_df)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,RF + DODA,5,Logistic Regression,"worst area, worst concave points, mean concave...",0.938596,0.985075,0.916667,0.949640,0.990741
1,RF + DODA,5,Random Forest,"worst area, worst concave points, mean concave...",0.929825,0.957143,0.930556,0.943662,0.986111
2,RF + DODA,5,XGBoost,"worst area, worst concave points, mean concave...",0.921053,0.943662,0.930556,0.937063,0.982474
3,RF + DODA,10,Logistic Regression,"worst area, worst concave points, mean concave...",0.912281,0.969697,0.888889,0.927536,0.991402
4,RF + DODA,10,Random Forest,"worst area, worst concave points, mean concave...",0.956140,0.958904,0.972222,0.965517,0.989087
5,RF + DODA,10,XGBoost,"worst area, worst concave points, mean concave...",0.964912,0.959459,0.986111,0.972603,0.991071
6,RF + DODA,15,Logistic Regression,"worst area, worst concave points, mean concave...",0.938596,0.957746,0.944444,0.951049,0.994048
7,RF + DODA,15,Random Forest,"worst area, worst concave points, mean concave...",0.956140,0.958904,0.972222,0.965517,0.994048
8,RF + DODA,15,XGBoost,"worst area, worst concave points, mean concave...",0.947368,0.958333,0.958333,0.958333,0.994378
9,RF + DODA,20,Logistic Regression,"worst area, worst concave points, mean concave...",0.956140,0.985507,0.944444,0.964539,0.994378


In [24]:
rf_doda_results_df.to_csv(
    "../../../results/breast_cancer/doda_results/rf_doda_results.csv",
    index=False
)

In [25]:
# =============================================================================
# COMBINED COMPARISON
# =============================================================================

comparison_df = pd.concat(
    [
        rf_results_df,
        rf_doda_results_df
    ],
    ignore_index=True
)


display(
    comparison_df
)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,Random Forest,5,Logistic Regression,"worst area, worst concave points, mean concave...",0.938596,0.985075,0.916667,0.949640,0.990741
1,Random Forest,5,Random Forest,"worst area, worst concave points, mean concave...",0.938596,0.971014,0.930556,0.950355,0.986772
2,Random Forest,5,XGBoost,"worst area, worst concave points, mean concave...",0.921053,0.943662,0.930556,0.937063,0.982474
3,Random Forest,10,Logistic Regression,"worst area, worst concave points, mean concave...",0.912281,0.969697,0.888889,0.927536,0.991402
4,Random Forest,10,Random Forest,"worst area, worst concave points, mean concave...",0.956140,0.958904,0.972222,0.965517,0.990079
5,Random Forest,10,XGBoost,"worst area, worst concave points, mean concave...",0.973684,0.972603,0.986111,0.979310,0.992394
6,Random Forest,15,Logistic Regression,"worst area, worst concave points, mean concave...",0.938596,0.957746,0.944444,0.951049,0.994048
7,Random Forest,15,Random Forest,"worst area, worst concave points, mean concave...",0.947368,0.958333,0.958333,0.958333,0.992725
8,Random Forest,15,XGBoost,"worst area, worst concave points, mean concave...",0.947368,0.958333,0.958333,0.958333,0.995040
9,Random Forest,20,Logistic Regression,"worst area, worst concave points, mean concave...",0.956140,0.985507,0.944444,0.964539,0.995701


In [26]:
for k in k_values:

    selector = rf_doda_results[k]["selector"]

    raw_rank = [
        x[0]
        for x in sorted(
            selector.raw_math_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    doda_rank = [
        x[0]
        for x in sorted(
            selector.final_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    print("\n" + "=" * 70)
    print(f"TOP {k}")
    print("=" * 70)

    print("Raw Math Ranking:")
    print(raw_rank)

    print("\nDODA Ranking:")
    print(doda_rank)

    print("\nTop-K Same:",
          raw_rank[:k] == doda_rank[:k])


TOP 5
Raw Math Ranking:
['worst area', 'worst concave points', 'mean concave points', 'worst perimeter', 'worst radius', 'mean concavity', 'mean perimeter', 'mean area', 'worst concavity', 'mean radius', 'area error', 'worst texture', 'radius error', 'worst compactness', 'mean compactness', 'mean texture', 'worst symmetry', 'worst smoothness', 'perimeter error', 'compactness error', 'worst fractal dimension', 'mean smoothness', 'texture error', 'smoothness error', 'fractal dimension error', 'mean symmetry', 'concave points error', 'symmetry error', 'concavity error', 'mean fractal dimension']

DODA Ranking:
['worst area', 'worst concave points', 'mean concave points', 'worst perimeter', 'worst radius', 'mean concavity', 'mean perimeter', 'worst concavity', 'mean area', 'mean radius', 'area error', 'worst texture', 'worst compactness', 'mean compactness', 'radius error', 'worst symmetry', 'mean texture', 'worst smoothness', 'perimeter error', 'mean smoothness', 'compactness error', 'wo

# 3. Random Forest + DODA: Experiment Interpretation

## 3.1 Overall Observation

The Random Forest-based feature-selection experiment produced strong predictive performance across all three downstream classifiers. Accuracy generally ranged from 0.912 to 0.965, recall from 0.889 to 0.986, F1-score from 0.928 to 0.972, and ROC-AUC remained high across configurations.

DODA did **not** produce a uniform improvement. Its effect depended on both the downstream model and the Top-K feature subset. In this experiment, DODA mainly refined the Random Forest-derived feature ranking; performance was preserved in several configurations, improved in selected Random Forest and XGBoost settings, and reduced for Logistic Regression at Top-K = 10.

## 3.2 Change in Feature Ranking

The main ranking change was not a wholesale replacement of Random Forest-important features. Instead, DODA adjusted the ordering by multiplying normalized mathematical importance by clinical weights.

| Feature | RF Mathematical Rank | DODA Rank | Rank Change | Clinical Weight | Interpretation |
|---------|---------------------:|----------:|------------:|----------------:|----------------|
| worst area | 1 | 1 | 0 | 1.0 | Retained as the highest-ranked feature. |
| worst concave points | 2 | 2 | 0 | 1.0 | Retained as a dominant feature. |
| worst radius | 4 | 3 | +1 | 1.0 | Promoted because its clinical weight preserved its full mathematical score. |
| mean concave points | 3 | 4 | -1 | 0.9 | Slightly demoted despite high mathematical importance because its clinical weight was lower than 1.0. |
| worst perimeter | 5 | 5 | 0 | 1.0 | Retained in the Top-5. |
| mean concavity | 6 | 6 | 0 | 0.9 | Retained as a highly ranked feature. |
| mean perimeter | 7 | 7 | 0 | 0.9 | Retained as a highly ranked feature. |
| mean radius | 8 | 8 | 0 | 0.9 | Retained in the Top-10. |
| worst concavity | 11 | 9 | +2 | 1.0 | Promoted into the Top-10 due to high clinical weighting. |
| mean area | 10 | 10 | 0 | 0.9 | Retained at the Top-10 boundary. |
| area error | 9 | 11 | -2 | 0.5 | Demoted below the Top-10 after clinical weighting. |
| worst compactness | 13 | 12 | +1 | 0.8 | Mildly promoted. |
| worst texture | 12 | 13 | -1 | 0.6 | Mildly demoted. |
| mean compactness | 15 | 14 | +1 | 0.7 | Mildly promoted within the Top-15. |
| radius error | 14 | 15 | -1 | 0.5 | Mildly demoted within the Top-15. |
| worst symmetry | 19 | 17 | +2 | 0.6 | Promoted within the Top-20. |
| mean symmetry | 24 | 21 | +3 | 0.5 | Promoted, but remained outside the Top-20. |
| texture error | 23 | 26 | -3 | 0.3 | Demoted among lower-ranked features. |
| fractal dimension error | 26 | 29 | -3 | 0.2 | Demoted among lower-ranked features. |

DODA preserved the strongest Random Forest mathematical features while promoting clinically weighted features such as worst radius, worst concavity, worst compactness, mean compactness, and worst symmetry. Lower clinical weights contributed to demotions such as area error, radius error, texture error, and fractal dimension error.

## 3.3 Top-K Feature Selection

### Top-K = 5

RF baseline features: worst area, worst concave points, mean concave points, worst radius, worst perimeter.

RF + DODA features: worst area, worst concave points, worst radius, mean concave points, worst perimeter.

The same five features were retained. The main change was internal ordering: worst radius moved ahead of mean concave points.

### Top-K = 10

RF baseline features: worst area, worst concave points, mean concave points, worst radius, worst perimeter, mean concavity, mean perimeter, mean radius, area error, mean area.

RF + DODA features: worst area, worst concave points, worst radius, mean concave points, worst perimeter, mean concavity, mean perimeter, mean radius, worst concavity, mean area.

The Top-10 set changed by one feature: worst concavity entered and area error left.

### Top-K = 15

RF baseline features: worst area, worst concave points, mean concave points, worst radius, worst perimeter, mean concavity, mean perimeter, mean radius, area error, mean area, worst concavity, worst texture, worst compactness, radius error, mean compactness.

RF + DODA features: worst area, worst concave points, worst radius, mean concave points, worst perimeter, mean concavity, mean perimeter, mean radius, worst concavity, mean area, area error, worst compactness, worst texture, mean compactness, radius error.

The same 15 features were retained, but their ranking order changed.

### Top-K = 20

RF baseline features: worst area, worst concave points, mean concave points, worst radius, worst perimeter, mean concavity, mean perimeter, mean radius, area error, mean area, worst concavity, worst texture, worst compactness, radius error, mean compactness, mean texture, worst smoothness, mean smoothness, worst symmetry, perimeter error.

RF + DODA features: worst area, worst concave points, worst radius, mean concave points, worst perimeter, mean concavity, mean perimeter, mean radius, worst concavity, mean area, area error, worst compactness, worst texture, mean compactness, radius error, mean texture, worst symmetry, worst smoothness, mean smoothness, perimeter error.

The same 20 features were retained. DODA mainly refined the ordering within the selected feature set.

| Top-K | Retained Features | Entered with DODA | Left after DODA | Interpretation |
|------:|-------------------|------------------|-----------------|----------------|
| 5 | worst area; worst concave points; mean concave points; worst radius; worst perimeter | None | None | Ranking refinement only. |
| 10 | 9 of 10 features retained | worst concavity | area error | Partial feature replacement at the Top-10 boundary. |
| 15 | All 15 features retained | None | None | Feature set preserved; ranking order changed. |
| 20 | All 20 features retained | None | None | Feature set preserved; ranking order changed. |

## 3.4 Top-K Stability Between RF and RF + DODA

The printed `Top-K Same: False` result reflects exact ordered-list equality, not feature-set equality. A Top-K list can have the same features but still return `False` if their internal order changes.

| Top-K | Common Features | Overlap | Stability | Main Change |
|------:|----------------:|--------:|-----------|-------------|
| 5 | 5 / 5 | 100.0% | Highly Stable | Feature set preserved; worst radius and mean concave points changed order. |
| 10 | 9 / 10 | 90.0% | Highly Stable | worst concavity entered; area error left. |
| 15 | 15 / 15 | 100.0% | Highly Stable | Feature set preserved; internal ordering changed. |
| 20 | 20 / 20 | 100.0% | Highly Stable | Feature set preserved; internal ordering changed. |

Ranking stability refers to whether features remain in the same order. Feature-set stability refers to whether the same features are selected regardless of order. In this experiment, ranking stability changed, but feature-set stability was high.

# 4. Predictive Performance Comparison

The following tables compare Random Forest baseline feature selection with RF + DODA feature selection using the executed notebook outputs. Values are shown as `baseline -> DODA`.

## 4.1 Logistic Regression Results

| Top-K | Accuracy | Precision | Recall | F1-score | ROC-AUC | Overall Effect |
|------:|----------|-----------|--------|----------|---------|----------------|
| 5 | 0.939 -> 0.939 | 0.985 -> 0.985 | 0.917 -> 0.917 | 0.950 -> 0.950 | 0.991 -> 0.991 | Maintained / Comparable |
| 10 | 0.947 -> 0.912 | 0.985 -> 0.970 | 0.931 -> 0.889 | 0.957 -> 0.928 | 0.992 -> 0.991 | Reduced |
| 15 | 0.939 -> 0.939 | 0.958 -> 0.958 | 0.944 -> 0.944 | 0.951 -> 0.951 | 0.994 -> 0.994 | Maintained / Comparable |
| 20 | 0.956 -> 0.956 | 0.986 -> 0.986 | 0.944 -> 0.944 | 0.965 -> 0.965 | 0.994 -> 0.994 | Maintained / Comparable |

The best Logistic Regression baseline configuration was Top-K = 20, with Accuracy = 0.956, Recall = 0.944, F1-score = 0.965, and ROC-AUC = 0.994. The best DODA configuration was also Top-K = 20 with the same values. DODA maintained comparable performance at Top-K = 5, 15, and 20, but reduced performance at Top-K = 10.

## 4.2 Random Forest Results

| Top-K | Accuracy | Precision | Recall | F1-score | ROC-AUC | Overall Effect |
|------:|----------|-----------|--------|----------|---------|----------------|
| 5 | 0.921 -> 0.921 | 0.944 -> 0.944 | 0.931 -> 0.931 | 0.937 -> 0.937 | 0.986 -> 0.984 | Maintained / Comparable |
| 10 | 0.956 -> 0.956 | 0.972 -> 0.959 | 0.958 -> 0.972 | 0.965 -> 0.966 | 0.989 -> 0.987 | Improved |
| 15 | 0.956 -> 0.965 | 0.959 -> 0.972 | 0.972 -> 0.972 | 0.966 -> 0.972 | 0.993 -> 0.994 | Improved |
| 20 | 0.956 -> 0.956 | 0.959 -> 0.959 | 0.972 -> 0.972 | 0.966 -> 0.966 | 0.993 -> 0.996 | Maintained / Comparable |

The best Random Forest baseline configurations were Top-K = 15 and Top-K = 20, both with Accuracy = 0.956, Recall = 0.972, and F1-score = 0.966; Top-K = 20 had the higher ROC-AUC. The best DODA Random Forest configuration was Top-K = 15, with Accuracy = 0.965, Recall = 0.972, F1-score = 0.972, and ROC-AUC = 0.994. DODA improved Random Forest performance most clearly at Top-K = 15.

## 4.3 XGBoost Results

| Top-K | Accuracy | Precision | Recall | F1-score | ROC-AUC | Overall Effect |
|------:|----------|-----------|--------|----------|---------|----------------|
| 5 | 0.921 -> 0.921 | 0.944 -> 0.944 | 0.931 -> 0.931 | 0.937 -> 0.937 | 0.982 -> 0.982 | Maintained / Comparable |
| 10 | 0.956 -> 0.965 | 0.959 -> 0.959 | 0.972 -> 0.986 | 0.966 -> 0.973 | 0.982 -> 0.991 | Improved |
| 15 | 0.947 -> 0.956 | 0.958 -> 0.959 | 0.958 -> 0.972 | 0.958 -> 0.966 | 0.995 -> 0.996 | Improved |
| 20 | 0.947 -> 0.947 | 0.946 -> 0.946 | 0.972 -> 0.972 | 0.959 -> 0.959 | 0.995 -> 0.997 | Maintained / Comparable |

The best XGBoost baseline configuration by F1-score was Top-K = 10. The best XGBoost DODA configuration was also Top-K = 10, with Accuracy = 0.965, Recall = 0.986, F1-score = 0.973, and ROC-AUC = 0.991. XGBoost appeared sensitive to the DODA ranking at Top-K = 10 and Top-K = 15, where both recall and F1-score improved.

## 4.4 Best Performing Configurations

| Category | Configuration | Accuracy | Recall | F1-score | ROC-AUC |
|----------|---------------|---------:|-------:|---------:|--------:|
| Best RF baseline configuration | Random Forest baseline, Random Forest, Top-K = 20 | 0.956 | 0.972 | 0.966 | 0.993 |
| Best RF + DODA configuration | RF + DODA, XGBoost, Top-K = 10 | 0.965 | 0.986 | 0.973 | 0.991 |
| Highest Accuracy | RF + DODA, Random Forest, Top-K = 15; RF + DODA, XGBoost, Top-K = 10 | 0.965 | 0.972 / 0.986 | 0.972 / 0.973 | 0.994 / 0.991 |
| Highest Recall | RF + DODA, XGBoost, Top-K = 10 | 0.965 | 0.986 | 0.973 | 0.991 |
| Highest F1-score | RF + DODA, XGBoost, Top-K = 10 | 0.965 | 0.986 | 0.973 | 0.991 |
| Highest ROC-AUC | RF + DODA, XGBoost, Top-K = 20 | 0.947 | 0.972 | 0.959 | 0.997 |

## 4.5 Importance of Accuracy, Recall, and F1-score

Accuracy is important because it summarizes the overall proportion of correctly classified samples. Recall is particularly important in this breast cancer classification setting because missed positive cases can be clinically serious. F1-score is useful because it balances Precision and Recall rather than emphasizing only one of them.

The strongest balance of these metrics was observed for RF + DODA with XGBoost at Top-K = 10, which achieved Accuracy = 0.965, Recall = 0.986, and F1-score = 0.973. RF + DODA with Random Forest at Top-K = 15 also showed a strong balance, with Accuracy = 0.965, Recall = 0.972, and F1-score = 0.972.

# 5. Effect of Clinical Knowledge

Clinical weighting affected the Random Forest ranking by reducing the scores of features with lower clinical weights and preserving the scores of features with weight 1.0. This promoted features such as worst radius, worst concavity, worst compactness, mean compactness, worst symmetry, and mean symmetry. It demoted features such as mean concave points, area error, radius error, worst texture, texture error, and fractal dimension error.

DODA preserved the strongest statistically important features: worst area, worst concave points, worst radius, mean concave points, worst perimeter, mean concavity, mean perimeter, and mean radius all remained highly ranked. The only Top-10 feature-set replacement was worst concavity entering in place of area error.

The results suggest that clinical knowledge mainly refined the ranking rather than substantially replacing the Random Forest feature set. However, a higher DODA rank should not be interpreted as proof that a feature is independently more predictive. The mathematical feature importance, clinical weighting, and downstream predictive performance are related but distinct components of the experiment.

Performance changes were not consistently positive. DODA improved selected Random Forest and XGBoost configurations, preserved several configurations, and reduced Logistic Regression performance at Top-K = 10.

# 6. Main Findings

| Finding | Observation |
|---------|-------------|
| Strongest RF baseline configuration | Random Forest with Random Forest Top-K = 20 achieved a strong baseline balance: Accuracy = 0.956, Recall = 0.972, F1-score = 0.966, and ROC-AUC = 0.993. |
| Strongest RF + DODA configuration | XGBoost with RF + DODA Top-K = 10 achieved Accuracy = 0.965, Recall = 0.986, F1-score = 0.973. |
| Highest Accuracy | 0.965, achieved by RF + DODA Random Forest Top-K = 15 and RF + DODA XGBoost Top-K = 10. |
| Highest Recall | 0.986, achieved by RF + DODA XGBoost Top-K = 10. |
| Highest F1-score | 0.973, achieved by RF + DODA XGBoost Top-K = 10. |
| Strongest DODA improvement | XGBoost Top-K = 10 improved Accuracy from 0.956 to 0.965, Recall from 0.972 to 0.986, and F1-score from 0.966 to 0.973. |
| Largest performance reduction | Logistic Regression Top-K = 10 declined from Accuracy = 0.947, Recall = 0.931, F1-score = 0.957 to Accuracy = 0.912, Recall = 0.889, F1-score = 0.928. |
| Top-K stability pattern | Feature-set overlap was 100% for Top-K = 5, 15, and 20, and 90% for Top-K = 10. |
| Main feature-ranking change | worst radius moved above mean concave points; worst concavity moved into the Top-10 while area error moved out. |
| Overall effect of clinical knowledge | Clinical knowledge produced meaningful ranking refinement, but its predictive effect was model- and Top-K-dependent. |

# 7. Research Interpretation

These results do not show that incorporating clinical knowledge through DODA consistently improves predictive performance. The evidence is more nuanced. DODA preserved performance in several configurations, improved selected Random Forest and XGBoost configurations, and reduced Logistic Regression performance at Top-K = 10.

The effect depended on both Top-K and downstream model. XGBoost showed the clearest gains at Top-K = 10 and Top-K = 15, while Random Forest improved most clearly at Top-K = 15. Logistic Regression was largely unchanged except for a reduction at Top-K = 10.

From a feature-selection perspective, DODA acted mainly as a ranking refinement method. It preserved most of the Random Forest-derived feature set while adjusting the internal order according to clinical weights. The experiment therefore supports the potential usefulness of clinical weighting, but only as a cautious, dataset-specific finding rather than evidence of general superiority.

# 8. Conclusion

RF + DODA did not consistently improve predictive performance across all models and Top-K values. It preserved performance in several configurations and improved selected Random Forest and XGBoost results, especially RF + DODA with XGBoost at Top-K = 10 and RF + DODA with Random Forest at Top-K = 15.

The feature ranking changed meaningfully, but the selected Top-K feature sets remained highly stable. The Top-5, Top-15, and Top-20 sets were fully preserved, while the Top-10 set changed by one feature. Clinical weighting therefore produced ranking refinement rather than a substantial replacement of the Random Forest-derived features.

The main limitation is that this experiment is based on one breast cancer dataset and one evaluation split, so it does not establish statistical significance or broader generalization. In this experiment, incorporating clinical knowledge through DODA primarily refined the Random Forest feature ranking, with predictive effects that were dependent on the model and Top-K setting rather than uniformly beneficial.